swarm.go

```diff
@@ -438,18 +434,10 @@ func (s *Swarm) Backoff() *DialBackoff {
 
 // notifyAll sends a signal to all Notifiees
 func (s *Swarm) notifyAll(notify func(inet.Notifiee)) {
-	var wg sync.WaitGroup
-
 	s.notifs.RLock()
-	wg.Add(len(s.notifs.m))
 	for f := range s.notifs.m {
-		go func(f inet.Notifiee) {
-			defer wg.Done()
-			notify(f)
-		}(f)
+		notify(f)
 	}
-
-	wg.Wait()
 	s.notifs.RUnlock()
```

Comment:
This is probably fine but we should profile this in an application (e.g., go-ipfs) as we've had a _ton_ of issues around notifications. Given that, I'm really not sure this change is worth it. Short-lived goroutines aren't _that_ expensive.

Reply:
Recently I saw bitswap lock up about 400 notifications, it will be better after the current rewrite.

---

Also I think if we leave the goroutine code there, we can move the `RUnlock()` above `wg.Wait()`

Reply:
How recently and are you saying that doing these in series (this patch) will make it better?

Reply:
Two days ago but it was on a 0.4.18.
No it won't, this node would have probably died if not for it. Bitswap should handle those notifications with priority (loop over select with default). I think it will be better now after the rewrite.

Reply:
Spawning unbounded goroutines with the assumption that "they'll get processed" is not a good idea. The propagation of notifications needs to be overhauled (better yet, reduced, there is an awful lot of abstraction inversion floating around).

Reply:
Do you intend to create an issue for this? Is there one in IPFS we can observe?

Reply:
As there was go-bitswap rewrite I will wait for it to reappear in future.